server/etcdserver/api/etcdhttp/health.go

```diff
@@ -193,3 +224,164 @@ func checkAPI(lg *zap.Logger, srv ServerHealth, serializable bool) Health {
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) InstallHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
+	// installs the http handler for the root path.
+	reg.installHttpEndpoint(lg, mux, reg.path, reg.checkNames...)
+	for _, checkName := range reg.checkNames {
+		// installs the http handler for the individual check sub path.
+		reg.installHttpEndpoint(lg, mux, path.Join(reg.path, checkName), checkName)
+	}
+}
+
+func (reg *CheckRegistry) runHealthChecks(ctx context.Context, checkNames ...string) (Health, error) {
+	h := Health{Health: "true"}
+	failedChecks := []string{}
+	var individualCheckOutput bytes.Buffer
+	for _, checkName := range checkNames {
+		check, found := reg.checks[checkName]
+		if !found {
+			return Health{Health: "false"}, fmt.Errorf("Health check: %s not registered", checkName)
+		}
+		if err := check(ctx); err != nil {
+			fmt.Fprintf(&individualCheckOutput, "[-]%s failed: %v\n", checkName, err)
+			failedChecks = append(failedChecks, checkName)
+		} else {
+			fmt.Fprintf(&individualCheckOutput, "[+]%s ok\n", checkName)
+		}
+	}
+	h.Reason = individualCheckOutput.String()
+	if len(failedChecks) > 0 {
+		h.Health = "false"
+	}
+	return h, nil
+}
+
+// installHttpEndpoint installs the http handler for the root path.
+func (reg *CheckRegistry) installHttpEndpoint(lg *zap.Logger, mux *http.ServeMux, path string, checks ...string) {
+	hfunc := func(r *http.Request) (Health, error) {
+		// extracts the health check names to be excludeList from the query param
+		excluded := getQuerySet(r, "exclude")
+		// extracts the health check names to be allowList from the query param
+		included := getQuerySet(r, "allowlist")
+
+		filteredCheckNames, err := filterCheckList(checks, included, excluded)
+		if err != nil {
+			return Health{Health: "false"}, err
+		}
+		return reg.runHealthChecks(r.Context(), filteredCheckNames...)
+	}
+	mux.Handle(path, newHealthHandler(path, lg, hfunc))
+}
+
+func filterCheckList(checks []string, included StringSet, excluded StringSet) (filteredList []string, err error) {
+	switch {
+	case len(excluded) > 0 && len(included) > 0:
+		return nil, fmt.Errorf("both an allowlist and an exclude list are specified in the query")
+	case len(excluded) > 0 && len(included) == 0:
+		return filterExcludedChecks(checks, excluded)
+	case len(excluded) == 0 && len(included) > 0:
+		return filterIncludedChecks(checks, included)
+	default:
+		return checks, nil
+	}
+}
+
+func filterExcludedChecks(checks []string, excluded StringSet) (filteredList []string, err error) {
+	for _, chk := range checks {
+		if _, found := excluded[chk]; found {
+			delete(excluded, chk)
+			continue
+		}
+		filteredList = append(filteredList, chk)
+	}
+	if len(excluded) > 0 {
+		return nil, fmt.Errorf("some health checks cannot be excluded: no matches for %s", formatQuoted(excluded.List()...))
+	}
+	return filteredList, nil
+}
+
+func filterIncludedChecks(checks []string, included StringSet) (filteredList []string, err error) {
```

Comment:
why do we need included checks?

Reply:
Allowlist could be a nice option for users that only care about a few selected checks and don't want be affected by newly added checks.

Reply:
After checking the old discussions in https://github.com/kubernetes/kubernetes/pull/70676#discussion_r231291895 and discussing with Chao, we decided to follow Kubernetes convention and drop the allowlist, and make exclude non-exist check not a failure. Thank you @logicalhan for bringing the context to our attention!